implementation/sequicity/tsd_net.py

```diff
@@ -270,6 +371,7 @@ def __init__(self, embed_size, hidden_size, vocab_size, degree_size, dropout_rat
         self.vocab = vocab
 
     def get_sparse_selective_input(self, x_input_np):
+        # BSpanDecoder uses get_sparse_input_aug. TODO: What is the difference between these two?
```

Comment:
## `get_sparse_input_aug` 
gets input with shape `[T,B]` where `T` is maximum length of sequence and `B` is batch (?). The input contains one-hot encoded sentences 
(eg. [[0,0,0], [3,2,1],[1,0,0]] is three sentences (for some reason, `t==0` is ignored), one of length 2 and the of length 1: word3 word1, word2, word1)

It converts this to array of shape `[B,T,aug_V]` where `aug_V` is of length `vocab_size+T`.
The result is sparse array.
So from previous example: 
```
[
  [[0,0,0,0,1,0,0], [0,0,1,0,0,0,0]], 
  [[0,0,0,1,0,0,0], [0,0,0,0,0,0,0]],
  [[0,0,1,0,0,0,0], [0,0,0,0,0,0,0]]
]
```
*where those zeros are actually 10^-9*
If there is a unknown symbol `x` at time `t` is adds a `1` at position `vocab_size+t` in the last (`aug_V`) dimension (don't know why - they probably want to distinguish unknown symbol at different positions).

## `get_sparse_selective_input`
does a similar thing. Input/output shapes are the same.
However, if the word in question is one of `reqs = ['address', 'phone', 'postcode', 'pricerange', 'area']`,
it appends `_SLOT` at the end of it (eg. `phone` -> `phone_SLOT`), and then it gets the corresponding sparse representation.

Also if word is unknown OR its one-hot index is larger than `vocab_size` then `5.0` is added to the corresponding position in sparse representation instead of `1.0` (don't know why - IHMO it should have the same effect).

------

**Note** that these two methods use different code to handle the ignoring of the first *time index*, but in my opinion they both do the same thing with respect to this.

Reply:
> and B is batch (?)
Yes, B is batch, see here: https://github.com/pixelneo/dialogue-transformer-e2e/blob/a036535205146105df7eb22d4661beb79186ac8c/implementation/sequicity/tsd_net.py#L196

>  (for some reason, t==0 is ignored)
Probably because outputs are shifted right by one, see any Transformer diagram.

> length vocab_size+T
Good job! I remember this from reading the paper. We thus have one-hot-encoded words on the output - and they can either be from the vocabulary or from the current input sentence. Being able to encode OOV words from the current input sentence makes copying possible.

If we want to just add in a vanilla transformer without copying, we should only use `vocab_size`, not `vocab_size + T`. However, once we add the copy mechanism, we need to use `vocab_size + T`.

> If there is a unknown symbol x at time t is adds a 1 at position vocab_size+t in the last (aug_V) dimension (don't know why - they probably want to distinguish unknown symbol at different positions).

This is how we encode an OOV word that is in the current input sentence. This way we can copy it to the output using a copy mechanism.

Reply:
> if word is unknown OR its one-hot index is larger than vocab_size

I don't understand why they have the OR here. (I am just thinking about it, I didn't read the code). Aren't the statements equivalent? That is word is unknown if and only if one-hot index > vocab_size?

> 5.0 is added to the corresponding position in sparse representation instead of 1.0 (don't know why - IHMO it should have the same effect).

Just guessing: I believe this is done to greatly increase the copy probability. It makes sence to me that OOV words and SLOTS should have a much higher chance to be copied. The authors actually added the copy probability mainly for these cases. (We should see if this 5.0 is in the paper.)

> Note that these two methods use different code to handle the ignoring of the first time index, but in my opinion they both do the same thing with respect to this.

Great news! Do you think we can just have one function instead of two to make the code simpler?

Reply:
Thank you so, much. Good job!  👍👏

 Let's rewrite this and commit this inside the code as comments. And resolve the PR conversation. After this, the TODO:

> Understand get_sparse_input_aug and get_sparse_selective_input. Do we need to use/re-implement them?

is complete. I further propose a new TODO:

- "Implement a `parse_input` function for our new Transformer Decoder that would unify the two original functions: `get_sparse_input_aug` and  `get_sparse_selective_input `. We can add in a (we are using a copy mechanism) on/off switch into the function."

Reply:
@pixelneo Pushed a commit with the comments. Can you double-check it and have a look at the TODOs?

Reply:
> I don't understand why they have the OR here. (I am just thinking about it, I didn't read the code). Aren't the statements equivalent? That is word is unknown if and only if one-hot index > vocab_size?
Maybe I did not understand it fully, either the word is `<unk>` which I thought is "unknown" or the one-hot index of the word is > vocab size (which could be the words to copy).

Reply:
> Do you think we can just have one function instead of two to make the code simpler?

I would not do that. They just handle the first index differently (with the same result) but otherwise they do different things. I'd leave unifying them for later.

Reply:
> a new TODO
> "Implement a parse_input function for our new Transformer Decoder that would unify the two original functions: get_sparse_input_aug and get_sparse_selective_input. We can add in a (we are using a copy mechanism) on/off switch into the function."

Is it necessary, now? Also, the two function do slightly different things, `get_sparse_selective_input` is called only in `ResponseDecoder`, dealing with the belief span.

Reply:
> Maybe I did not understand it fully, either the word is <unk> which I thought is "unknown" or the one-hot index of the word is > vocab size (which could be the words to copy).

You are right. What I said didn't make sense.

Here is what I am trying to get straight in my head, actually:

1) `get_sparse_input_aug`
    - `u_input_np` and `prev_z_input_np`, both in `BSpanDecoder`.
    - Which sentences does this encode? User utterences and Bspans?
    - Input: `u_input_np` Can this get only vocabulary words and OOV words on the input?
    - Input: `prev_z_input_np` I guess 
    - And that it can only have vocabulary words and "copy words" on the output.
2) `get_sparse_selective_input`. 
    - called on `z_input_np` in `ResponseDecoder`
    - Which sentences does this encode? Responses? Bspans?
    - Input: SLOT values, vocabulary words, OOV words, "copy words" copied from the User utterance.
    - When does this get  an OOV word on the input?

I still don't know what we are converting exactly...

Reply:
> get_sparse_input_aug
> Which sentences does this encode? User utterences and Bspans?

Both, I guess. Certainly user inputs.
But I think also BSpans (`prev_z_input_np`) (difference from `get_sparse_selective_input` is for example that this does not create new slots)

> get_sparse_selective_input.
> Which sentences does this encode? Responses? Bspans?

BSpans, adding the `_SLOT` when necessary. 
Only then, from these outputs, responses are generated, replacing the `_SLOT`ed words with some KB data.